src/create-glamorous.js

```diff
@@ -0,0 +1,162 @@
+/*
+ * This is a relatively small abstraction that's ripe for open sourcing.
+ * Documentation is in the README.md
+ */
+import React, {Component} from 'react'
+import {PropTypes} from './react-compat'
+import {CHANNEL} from './constants'
+import getGlamorClassName from './get-glamor-classname'
+
+export default function createGlamorous(splitProps) {
+  /**
+   * This is the main export and the function that people
+   * interact with most directly.
+   *
+   * It accepts a component which can be a string or
+   * a React Component and returns
+   * a "glamorousComponentFactory"
+   * @param {String|ReactComponent} comp the component to render
+   * @param {Object} options helpful info for the GlamorousComponents
+   * @return {Function} the glamorousComponentFactory
+   */
+  return function glamorous(
+    comp,
+    {rootEl, displayName, forwardProps = []} = {},
+  ) {
+    return glamorousComponentFactory
+
+    /**
+     * This returns a React Component that renders the comp (closure)
+     * with a className based on the given glamor styles object(s)
+     * @param {...Object|Function} styles the styles to create with glamor.
+     *   If any of these are functions, they are invoked with the component
+     *   props and the return value is used.
+     * @return {ReactComponent} the ReactComponent function
+     */
+    function glamorousComponentFactory(...styles) {
+      /**
+       * This is a component which will render the comp (closure)
+       * with the glamorous styles (closure). Forwards any valid
+       * props to the underlying component.
+       * @param {Object} theme the theme object
+       * @return {ReactElement} React.createElement
+       */
+      class GlamorousComponent extends Component {
+        state = {theme: null}
+        setTheme = theme => this.setState({theme})
+
+        componentWillMount() {
+          const {theme} = this.props
+          if (this.context[CHANNEL]) {
+            // if a theme is provided via props,
+            // it takes precedence over context
+            this.setTheme(theme ? theme : this.context[CHANNEL].getState())
+          } else {
+            this.setTheme(theme || {})
+          }
+        }
+
+        componentWillReceiveProps(nextProps) {
+          if (this.props.theme !== nextProps.theme) {
+            this.setTheme(nextProps.theme)
+          }
+        }
+
+        componentDidMount() {
+          if (this.context[CHANNEL] && !this.props.theme) {
+            // subscribe to future theme changes
+            this.unsubscribe = this.context[CHANNEL].subscribe(this.setTheme)
+          }
+        }
+
+        componentWillUnmount() {
+          // cleanup subscription
+          this.unsubscribe && this.unsubscribe()
+        }
+
+        render() {
+          // in this function, we're willing to sacrafice a little on
```

Comment:
Thanks for the clarification! I now realize this comment is more about describing what this function is aiming for. I was trying to find out/understand where exactly readability is sacrificed for performance 😄 .

I do think it may be useful to point those trade-offs out if it's not obvious because further contributors may refactor some code without realizing what it's good for performance or feel scared to touch the code because it may be done for performance. 

But either way, I think the way it currently is is fine. 👍